### PR TITLE
CL/HIER: 2step reduce algorithm

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Moving last piece from https://github.com/openucx/ucc/pull/621
Originally implemented by @vspetrov

### CODE REVIEW

File: src/components/cl/hier/reduce/reduce_2step.c [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    size_t                count     = (rank == root) ?
+                                      args.args.dst.info.count :
+                                      args.args.src.info.count;
+    ucc_cl_hier_schedule_t *cl_schedule;
+    ucc_schedule_t         *schedule;
+    ucc_status_t            status;
+    int                     n_tasks, i, first_task, root_on_local_node;
+
+    root_on_local_node = ucc_team_ranks_on_same_node(root, rank, core_team);
+    n_tasks        = 0;
```
Comment (samnordmann) [inline src/components/cl/hier/reduce/reduce_2step.c:87]: align the "=" ?

File: src/components/cl/hier/reduce/reduce_2step.c [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        if (ucc_unlikely(UCC_OK != status)) {
+            goto out;
+        }
+        n_tasks++;
+    }
+
+    if (SBGP_ENABLED(cl_team, NODE_LEADERS)) {
+        if (n_tasks == 1) {
+            if (root != rank) {
+                /* args.args.dst.info.buffer = cl_schedule->scratch->addr; */
```
Comment (samnordmann) [inline src/components/cl/hier/reduce/reduce_2step.c:139]: remove the comment ?

**Review by samnordmann — APPROVED** [review]

**Comment by samnordmann** [comment]
Can you maybe add this algo in the tests ?

**Comment by Sergei-Lebedev** [comment]
> Can you maybe add this algo in the tests ?

added

File: src/components/cl/hier/cl_hier_coll.c [hunk 3]
```diff
@@ -22,14 +23,16 @@ ucc_status_t ucc_cl_hier_coll_init(ucc_base_coll_args_t *coll_args,
...
     case UCC_COLL_TYPE_ALLTOALL:
         return ucc_cl_hier_alltoall_init(coll_args, team, task);
     case UCC_COLL_TYPE_ALLTOALLV:
         return ucc_cl_hier_alltoallv_init(coll_args, team, task);
+    case UCC_COLL_TYPE_BARRIER:
+        return ucc_cl_hier_barrier_init(coll_args, team, task);
     case UCC_COLL_TYPE_BCAST:
         return ucc_cl_hier_bcast_2step_init(coll_args, team, task);
+    case UCC_COLL_TYPE_REDUCE:
+        return ucc_cl_hier_reduce_2step_init(coll_args, team, task);
```
Comment (bureddy) [inline src/components/cl/hier/cl_hier_coll.c:35]: qq, why do we call specific alg init func from CL init func()?  is this default?

**Review by bureddy — APPROVED** [review]

File: src/components/cl/hier/cl_hier_coll.c (hunk 3 above)
Comment (Sergei-Lebedev) [reply src/components/cl/hier/cl_hier_coll.c:35]: yes, it's actually the only reduce implementation in CL HIER. However it calls sbgp reduces which can come from different TLs.

**Pull request merged**